### ISSUE TITLE
System.InvalidOperationException: On Xamarin.Android, you have to specify the current Activity from which the browser pop-up will be displayed using the WithParentActivityOrWindow method.

### ISSUE THREAD

**Issue by hverma11** [Azure-Samples/active-directory-xamarin-native-v2#73]
On executing this line 
 authResult = await App.PCA.AcquireTokenInteractive(App.Scopes)
                                                      .WithParentActivityOrWindow(App.ParentWindow)
                                                      .ExecuteAsync();

It is throwing Exception:

05-22 14:32:40.165 E/mono    (29592): 
05-22 14:32:40.165 E/mono    (29592): Unhandled Exception:
05-22 14:32:40.165 E/mono    (29592): System.InvalidOperationException: On Xamarin.Android, you have to specify the current Activity from which the browser pop-up will be displayed using the WithParentActivityOrWindow method.
05-22 14:32:40.165 E/mono    (29592):   at Microsoft.Identity.Client.AcquireTokenInteractiveParameterBuilder.Validate () [0x00018] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.165 E/mono    (29592):   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1[T].ValidateAndCalculateApiId () [0x00000] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.165 E/mono    (29592):   at Microsoft.Identity.Client.AbstractPublicClientAcquireTokenParameterBuilder`1[T].ExecuteAsync (System.Threading.CancellationToken cancellationToken) [0x00000] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.165 E/mono    (29592):   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1[T].ExecuteAsync () [0x00006] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.165 E/mono    (29592):   at Variance.ViewModels.MainPageViewModel+<OnSignInSignOut>d__12.MoveNext () [0x00285] in C:\Mobility\test\Variation\App\Variance\Variance\ViewModels\MainPageViewModel.cs:78 
05-22 14:32:40.165 E/mono    (29592): --- End of stack trace from previous location where exception was thrown ---
05-22 14:32:40.165 E/mono    (29592):   at (wrapper dynamic-method) System.Object.24(intptr,intptr)
05-22 14:32:40.165 E/mono    (29592):   at (wrapper native-to-managed) System.Object.24(intptr,intptr)
05-22 14:32:40.172 E/mono-rt (29592): [ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: On Xamarin.Android, you have to specify the current Activity from which the browser pop-up will be displayed using the WithParentActivityOrWindow method.
05-22 14:32:40.172 E/mono-rt (29592):   at Microsoft.Identity.Client.AcquireTokenInteractiveParameterBuilder.Validate () [0x00018] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.172 E/mono-rt (29592):   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1[T].ValidateAndCalculateApiId () [0x00000] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.172 E/mono-rt (29592):   at Microsoft.Identity.Client.AbstractPublicClientAcquireTokenParameterBuilder`1[T].ExecuteAsync (System.Threading.CancellationToken cancellationToken) [0x00000] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.172 E/mono-rt (29592):   at Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder`1[T].ExecuteAsync () [0x00006] in <48941809037546559a7e7308bdd45147>:0 
05-22 14:32:40.172 E/mono-rt (29592):   at Variance.ViewModels.MainPageViewModel+<OnSignInSignOut>d__12.MoveNext () [0x00285] in C:\Mobility\test\Variation\App\Variance\Variance\ViewModels\MainPageViewModel.cs:78 
05-22 14:32:40.172 E/mono-rt (29592): --- End of stack trace from previous location where exception was thrown ---
05-22 14:32:40.172 E/mono-rt (29592):   at (wrapper dynamic-method) System.Object.24(intptr,intptr)
05-22 14:32:40.172 E/mono-rt (29592):   at (wrapper native-to-managed) System.Object.24(intptr,intptr)

**Comment by hverma11** [comment]
More information:

 I have already set the Android ParentWindow settings and referring same in MSAL authentication

 public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {   
        protected override void OnCreate(Bundle bundle)
        {
            System.Net.ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };

            TabLayoutResource = Resource.Layout.tabs;
            ToolbarResource = Resource.Layout.toolbar;
            base.SetTheme(Resource.Style.MyTheme);
            base.OnCreate(bundle);
                
            global::Xamarin.Forms.Forms.Init(this, bundle);
            UserDialogs.Init(this);
            ConfigurationManager.Initialize(new AndroidConfigurationStreamProviderFactory(() => this));

            LoadApplication(new App(new AndroidInitializer()));

            App.ParentWindow = this;
        }

       protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
        {
            base.OnActivityResult(requestCode, resultCode, data);
            AuthenticationContinuationHelper.SetAuthenticationContinuationEventArgs(requestCode, resultCode, data);

        }
    }

**Comment by jennyf19** [comment]
@hverma11 have you tried running this sample as-is on an Android device and see if it works for you?

**Comment by jennyf19** [comment]
@hverma11 we have made updates to the parent window...have you tried w/the latest version?

**Comment by robiharid** [comment]
> More information:
> 
> I have already set the Android ParentWindow settings and referring same in MSAL authentication
> 
> public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
> {
> protected override void OnCreate(Bundle bundle)
> {
> System.Net.ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };
> 
> ```
>         TabLayoutResource = Resource.Layout.tabs;
>         ToolbarResource = Resource.Layout.toolbar;
>         base.SetTheme(Resource.Style.MyTheme);
>         base.OnCreate(bundle);
>             
>         global::Xamarin.Forms.Forms.Init(this, bundle);
>         UserDialogs.Init(this);
>         ConfigurationManager.Initialize(new AndroidConfigurationStreamProviderFactory(() => this));
> 
>         LoadApplication(new App(new AndroidInitializer()));
> 
>         App.ParentWindow = this;
>     }
> 
>    protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
>     {
>         base.OnActivityResult(requestCode, resultCode, data);
>         AuthenticationContinuationHelper.SetAuthenticationContinuationEventArgs(requestCode, resultCode, data);
> 
>     }
> }
> ```

did you solve it? having same issue.

**Comment by jennyf19** [comment]
@robiharid I just tried out the sample and did not have the issue reported here. I have [pushed a branch](https://github.com/Azure-Samples/active-directory-xamarin-native-v2/tree/jennyf/webview) which breaks apart some of the logic, hopefully get some more info into the root cause.

Did you update the [manifest](https://github.com/Azure-Samples/active-directory-xamarin-native-v2/blob/jennyf/webview/UserDetailsClient/UserDetailsClient.Droid/Properties/AndroidManifest.xml#L12) to have your clientid?

**Comment by robiharid** [comment]
> Did you update the [manifest](https://github.com/Azure-Samples/active-directory-xamarin-native-v2/blob/jennyf/webview/UserDetailsClient/UserDetailsClient.Droid/Properties/AndroidManifest.xml#L12) to have your clientid?

Thanks for helping Jenny. I have built the app which works fine on my UWP but not so on the Android... I get the same ParentWndow error as the OP when the auth fails to acquire tokens silently it calls the acquiretokeninteractive which fails with the parentwindow being the cause.

Not sure why this is happening as it works on my UWP.

**Comment by jennyf19** [comment]
@robiharid interesting. We test w/a variety of devices and I don't think we've run into this issue. Let me try again today w/a tablet (any specific type?)
Also, can you send me the code you use for creating the parent window and the public client? 
Or, if you want to email me a repro, I can try that out as well -> jeferrie@microsoft.com

**Comment by jmprieur** [comment]
@jennyf19 : did you receive the info?

**Comment by jennyf19** [comment]
@jmprieur yes, but we don't have the same device @robiharid  is using, so not sure how to test this.

**Comment by jmprieur** [comment]
@jennyf19 : do we still want to keep this open? A lot has changed since.

**Comment by robiharid** [comment]
> @jennyf19 : do we still want to keep this open? A lot has changed since.

no it is fine, please close

**Comment by jmprieur** [comment]
Thanks for confirming, @robiharid. Closing.
cc: @jennyf19

**Comment by jennyf19** [comment]
thanks @robiharid

**Comment by lorne-olo** [comment]
> @hverma11 we have made updates to the parent window...have you tried w/the latest version?

> @jennyf19 : do we still want to keep this open? A lot has changed since.

I wish the docs were up-to-date. I'm having similar problems. The sample runs on my Android phone. My own app doesn't. Some of the doc instructions don't match the sample code (e.g. App.ParentActivityOrWindow = this), it's difficult to keep track of what I've done wrong. I'll keep checking.... I'm also having a problem regarding the keychain on iOS--exception says I don't have keychain access groups enabled--but I do.

**Comment by jennyf19** [comment]
@lorne-olo iOS can be a bit challenging to set up correctly, especially with Xamarin. Make sure you've changed all the bundleIds in the `Info.plist` to the values of your app, for example:
```csharp
<key>CFBundleURLName</key>
			<string>com.yourcompany.UserDetailsClient</string>
			<key>CFBundleURLSchemes</key>
			<array>
				<string>msala7d8cef0-4145-49b2-a91d-95c54051fa3f</string>
			</array>
```
You'll need to use your app's bundleid and clientId `<string>msal{your_client_id}</string>`

And your `Entitlements.plist` should have this:
```csharp
<dict>
	<key>keychain-access-groups</key>
	<array>
		<string>$(AppIdentifierPrefix)com.microsoft.adalcache</string>
	</array>
</dict>
```

For Android, what exactly are you seeing?

**Comment by lorne-olo** [comment]
@jennyf19  My problem with Android was me missing CrossCurrentActivity.Current.Init(...) -- it was plain as day in the sample code, but the doc was different. Problem solved!

My issue now is with iOS. My build **works fine** when done via **App Center** and uploaded to **TestFlight**. However, I guess due to provisioning files or something, if I do a debug build and run it on an iPhone connected to my Windows 10 box, I have a keychain problem where it says I **don't have keychain access groups enabled**. Again--this is just when I attempt to run it in debug mode on an **attached physical iPhone (without a Mac)**.

**Comment by jennyf19** [comment]
@lorne-olo glad you got android sorted out. @aiwangmicrosoft Can you help make sure the docs/sample code are the same? thank you.

For iOS, do you have all the [provisioning profile information](https://docs.microsoft.com/en-us/xamarin/ios/get-started/installation/device-provisioning/manual-provisioning?tabs=macos#provision-an-ios-device-for-development) set up? I think you have to have it connected to a Mac to get it to work and provision correctly.

**Comment by acaliaro** [comment]
Hi. I have the same problem with maui. Where can I find a sample? all links are broken

**Comment by AnithaRaniT** [comment]
Hi I'm also facing the same issue with MAUI, can someone help

**Comment by pave-xx** [comment]
I'm also facing the same issue with MAUI. @AnithaRaniT Have you found any solution?
